src/io/save_gals_hdf5.c

```diff
@@ -0,0 +1,1149 @@
+    master_file_id = H5Fcreate(master_fname, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+    CHECK_STATUS_AND_RETURN_ON_FAIL(master_file_id, FILE_NOT_FOUND,
+                                    "Can't open file %s for master file writing.\n", master_fname);
+
+    // We will keep track of how many galaxies were saved across all files per snapshot.
+    int64_t *ngals_allfiles_snap;
+    ngals_allfiles_snap = calloc(run_params->MAXSNAPS, sizeof(*(ngals_allfiles_snap)));
+
+    // The master file will be accessed as (e.g.,) f["Core0"]["Snap_63"]["StellarMass"].
+    // Hence we want to store the external links to the **root** group (i.e., "/"). 
+    root_group_id = H5Gopen2(master_file_id, "/", H5P_DEFAULT);
+    CHECK_STATUS_AND_RETURN_ON_FAIL(root_group_id, (int32_t) root_group_id,
+                                    "Failed to open the root group for the master file.\nThe file ID was %d\n",
+                                    (int32_t) master_file_id);
+
+    // At this point, all the files of all other processors have been created. So iterate over the
+    // number of processors and create links within this master file to those files.
+    char target_fname[3*MAX_STRING_LEN];
+    char core_fname[MAX_STRING_LEN];
+
+    for(int32_t task_idx = 0; task_idx < run_params->NTasks; ++task_idx) {
+
+        snprintf(core_fname, MAX_STRING_LEN - 1, "Core_%d", task_idx);
+        snprintf(target_fname, 3*MAX_STRING_LEN - 1, "./%s_%d.hdf5", run_params->FileNameGalaxies, task_idx);
+
+        // Make a symlink to the root of the target file.
+        status = H5Lcreate_external(target_fname, "/", root_group_id, core_fname, H5P_DEFAULT, H5P_DEFAULT);
+        CHECK_STATUS_AND_RETURN_ON_FAIL(status, (int32_t) status,
+                                        "Failed to create an external link to file %s from the master file.\n" 
+                                        "The group ID was %d and the group name was %s\n", target_fname,
+                                        (int32_t) root_group_id, core_fname);
+    }
+
+    // We've finished with the linking. Now let's create some attributes and datasets inside the header group.
+    group_id = H5Gcreate2(master_file_id, "Header", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+    CHECK_STATUS_AND_RETURN_ON_FAIL(group_id, (int32_t) group_id,
+                                    "Failed to create the Header group for the master file.\nThe file ID was %d\n",
+                                    (int32_t) master_file_id);
+
+    // When we're writing the header attributes for the master file, we don't have knowledge of trees.
+    // So pass a NULL pointer here instead of `forest_info`.
+    status = write_header_attributes(master_file_id, NULL, run_params);
+    if(status != EXIT_SUCCESS) {
+        return status;
+    }
+
+    // Redshift at each snapshot.
+    hsize_t dims[1];
+    dims[0] = run_params->MAXSNAPS;
+    float *ZZ_snapshot;
+    ZZ_snapshot = calloc(run_params->MAXSNAPS, sizeof(*(ZZ_snapshot)));
```

Comment:
As before - you should be able to write out the array directly through `run_params`. Since this pattern is repeating, you might want to capture this pattern into a function (`write_header()` ?)